### Oluwapelumi Fafowora (owf1)

In [1]:
from sklearn.utils import murmurhash3_32

In [30]:
sed = 13

### Implement a simple MinHash generator. Given an input string and m. Return m hashcodes

In [36]:
def minhash(stg, sd):
    '''
    Inputs a string stg
    Returns the minimum hash value using the 3-gram representation of the strings
    '''
    
    #Converting the stg into a set of 3-grams
    slen = len(stg)
    three_gram_lst = [stg[i:i+3] for i in range(slen-2)]    
    #hashes each 3-gram and returns the minhash value
    return min([murmurhash3_32(strg,seed = sd) for strg in three_gram_lst])
    

In [40]:
print(minhash("you up?",7))

-740032803


In [38]:
def minhash_generator(stg, m):
    '''
    Inputs a string stg and an integer m 
    Returns m minhashes of stg
    '''
    hash_arr = [minhash(stg, hsh) for hsh in range(m)]
    return hash_arr
    

In [42]:
print(minhash_generator("you up?",8))

[-31670852, -1884678989, -2051367432, -1588244665, -1630555212, -1852347288, -777814675, -740032803]


### Generate 100 MinHash values of the following two strings:

In [43]:
s1 = "The mission statement of the WCSCC and area employers recognize the importance of good attendance on the job. Any student whose absences exceed 18 days is jeopardizing their opportunity for advanced placement as well as hindering his/her likelihood for successfully completing their program."
s2 = "The WCSCC’s mission statement and surrounding employers recognize the importance of great attendance. Any student who is absent more than 18 days will loose the opportunity for successfully completing their trade program."

In [88]:
s1_minh_lst = minhash_generator(s1, 100)

In [89]:
s2_minh_lst = minhash_generator(s2, 100)

In [48]:
def compute_num_collisions(arr1,arr2):
    '''
    Inputs two arrays of equal sizes
    Returns the number of values that were the same at the same index divided by the len of the array
    '''
    len_arr = len(arr1)
    num_coll = 0
    for i in range(len_arr):
        if arr1[i] == arr2[i]:
            num_coll += 1
    return (num_coll/len_arr)

In [94]:
#estimating the jaccard similarity of s1 and s2
print(compute_num_collisions(s1_minh_lst, s2_minh_lst))
#0.45

0.45


In [76]:
def jaccard_sim(str1, str2):
    '''
    Inputs two strings s1 and s2
    Returns the  true jaccard similarity of the two strings
    '''
    set_str1 = set([])
    set_str2 = set([])
    
    #Creating the three gram representation of each of the strings
    #for string 1
    for i in range(len(str1)-2):
        set_str1.add(str1[i:i+3])
    #for string 2
    for i in range(len(str1)-2):
        set_str2.add(str2[i:i+3])
    
    #calculate the cardinality of the intersection and the union of the two
    card_intersection = set_str1.intersection(set_str2)
    card_union = set_str1.union(set_str2)
    return(len(card_intersection)/len(card_union))

In [95]:
jaccard_sim(s1,s2)
#0.5173611111111112

0.5173611111111112

My estimate using minhashes is 0.45 while the true jaccard similarity is about 0.517